src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.kafka.listener;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.util.Assert;
+
+import reactor.Environment;
+import reactor.core.processor.RingBufferProcessor;
+import reactor.fn.BiFunction;
+import reactor.fn.Consumer;
+import reactor.fn.Function;
+import reactor.fn.tuple.Tuple2;
+import reactor.rx.Stream;
+import reactor.rx.Streams;
+import reactor.rx.stream.GroupedStream;
+
+/**
+ * An {@link OffsetManager} that aggregates writes over a time or count window, using an underlying delegate to
+ * do the actual operations. Its purpose is to reduce the performance impact of writing operations
+ * wherever this is desirable.
+ * <p>
+ * Either a time window or a number of writes can be specified, but not both.
```

Comment:
This doesn't seem to be true anymore - the windowing operation in Reactor takes both count and time. The mutual exclusion was unique to Rx.